src/h3lib/lib/polyfill.c

```diff
@@ -0,0 +1,422 @@
+ * will be set and all iteration will return H3_NULL. It is the responsibility
+ * of the caller to check the error property after initialization.
+ *
+ * At any point in the iteration, starting once the struct is initialized, the
+ * output value can be accessed through the `cell` property.
+ *
+ * Note that initializing the iterator allocates memory. If an iterator is
+ * exhausted or returns an error that memory is released; otherwise it must be
+ * released manually with iterDestroyPolygon.
+ *
+ * @param  polygon Polygon to fill with cells
+ * @param  res     Resolution for output cells
+ * @param  flags   Bit mask of option flags
+ * @return         Initialized iterator, with the first value available
+ */
+IterCellsPolygon iterInitPolygon(const GeoPolygon *polygon, int res,
+                                 uint32_t flags) {
+    // Create the sub-iterator for compact cells
+    IterCellsPolygonCompact cellIter =
+        iterInitPolygonCompact(polygon, res, flags);
+    // Create the sub-iterator for children
+    IterCellsChildren childIter = iterInitParent(cellIter.cell, res);
+
+    IterCellsPolygon iter = {.cell = childIter.h,
+                             .error = cellIter.error,
+                             ._cellIter = cellIter,
+                             ._childIter = childIter};
+    return iter;
+}
+
+/**
+ * Increment the polyfill iterator, outputting the latest cell at the
+ * desired resolution.
+ *
+ * @param  iter Iterator to increment
+ */
+void iterStepPolygon(IterCellsPolygon *iter) {
+    if (iter->cell == H3_NULL) return;
+
+    // See if there are more children to output
+    iterStepChild(&(iter->_childIter));
+    if (iter->_childIter.h) {
+        iter->cell = iter->_childIter.h;
+        return;
+    }
+
+    // Otherwise, increment the polyfill iterator
+    iterStepPolygonCompact(&(iter->_cellIter));
+    if (iter->_cellIter.cell) {
+        _iterInitParent(iter->_cellIter.cell, iter->_cellIter._res,
+                        &(iter->_childIter));
+        iter->cell = iter->_childIter.h;
+        return;
+    }
+
+    // All done, set to null and report errors if any
+    iter->cell = H3_NULL;
+    iter->error = iter->_cellIter.error;
+}
+
+/**
+ * Destroy an iterator, releasing any allocated memory. Iterators destroyed in
+ * this manner are safe to use but will always return H3_NULL.
+ * @param  iter Iterator to destroy
+ */
+void iterDestroyPolygon(IterCellsPolygon *iter) {
+    iterDestroyPolygonCompact(&(iter->_cellIter));
+    iter->cell = H3_NULL;
+    iter->error = E_SUCCESS;
+}
+
+/**
+ * Parity implementation for polygonToCells
+ */
+H3Error H3_EXPORT(polygonToCells2)(const GeoPolygon *polygon, int res,
+                                   uint32_t flags, H3Index *out) {
+    IterCellsPolygon iter = iterInitPolygon(polygon, res, flags);
+    int64_t i = 0;
+    for (; iter.cell; iterStepPolygon(&iter)) {
+        out[i++] = iter.cell;
+    }
```

Comment:
you could also assert we don't go over the estimation size here (which would be an out of bounds write)

Reply:
I don't think we'd want this in production code, since we *assume* the caller has sized the memory correctly, but the `*out` array could be any size, so our assertion would only be useful as an internal check that we were meeting our contract. But I think can use the assertion macros to check this in tests.